From e5fa8fb567ed8618a23db79cc57402be8255fc5f Mon Sep 17 00:00:00 2001
From: Chew <chew@chew.pw>
Date: Wed, 30 Sep 2020 15:02:57 -0500
Subject: [PATCH] Version/Name changes


diff --git a/build.gradle b/build.gradle
index b835ea5..d7b5d57 100644
--- a/build.gradle
+++ b/build.gradle
@@ -22,7 +22,7 @@ plugins {
     id 'com.jfrog.bintray' version '1.8.4'
 }
 
-def versionInfo = [major: '3', minor: '0', revision: '5']
+def versionInfo = [major: '1', minor: '0', revision: '0']
 
 task build {
     group = 'build'
@@ -33,8 +33,8 @@ allprojects {
     apply plugin: 'com.jfrog.bintray'
     apply plugin: 'com.github.johnrengelman.shadow'
 
-    group = 'com.jagrosh'
-    version = versionInfo.values().join('.')
+    group = 'pw.chew'
+    version = versionInfo.values().join('.') + "_" + System.getenv("BUILD_NUMBER")
 
     ext {
         jdaVersion = '4.2.0_214'
@@ -110,10 +110,10 @@ allprojects {
         publications = ['BintrayRelease']
         publish = true
         pkg {
-            repo = 'maven'
-            name = 'JDA-Utilities'
+            repo = 'projects'
+            name = 'JDA-Chewtils'
             licenses = ['Apache-2.0']
-            vcsUrl = 'https://github.com/JDA-Applications/JDA-Utilities'
+            vcsUrl = 'https://github.com/Chew/JDA-Chewtils'
             version {
                 name = project.version
                 released = new Date()
diff --git a/settings.gradle b/settings.gradle
index 869ba6a..5783c90 100644
--- a/settings.gradle
+++ b/settings.gradle
@@ -13,7 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-rootProject.name = 'JDA-Utilities'
+rootProject.name = 'JDA-Chewtils'
 
 include ':command'
 include ':commons'
-- 
2.29.2

