From faae0ff447c9d2688d75a31ae2af55a4bd9ff754 Mon Sep 17 00:00:00 2001
From: Chew <chew@chew.pw>
Date: Wed, 30 Sep 2020 15:02:57 -0500
Subject: [PATCH] Version/Name changes


diff --git a/build.gradle b/build.gradle
index b835ea5..9482bf5 100644
--- a/build.gradle
+++ b/build.gradle
@@ -22,7 +22,7 @@ plugins {
     id 'com.jfrog.bintray' version '1.8.4'
 }
 
-def versionInfo = [major: '3', minor: '0', revision: '5']
+def versionInfo = [major: '1', minor: '0', revision: '0']
 
 task build {
     group = 'build'
@@ -33,7 +33,7 @@ allprojects {
     apply plugin: 'com.jfrog.bintray'
     apply plugin: 'com.github.johnrengelman.shadow'
 
-    group = 'com.jagrosh'
+    group = 'pw.chew'
     version = versionInfo.values().join('.')
 
     ext {
diff --git a/settings.gradle b/settings.gradle
index 869ba6a..5783c90 100644
--- a/settings.gradle
+++ b/settings.gradle
@@ -13,7 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-rootProject.name = 'JDA-Utilities'
+rootProject.name = 'JDA-Chewtils'
 
 include ':command'
 include ':commons'
-- 
2.23.0

