From 27720433916a328eb3ff33c312260d5b071a54f4 Mon Sep 17 00:00:00 2001
From: Chew <chew@chew.pw>
Date: Tue, 27 Oct 2020 19:27:20 -0500
Subject: [PATCH] Check user perms before bot perms


diff --git a/command/src/main/java/com/jagrosh/jdautilities/command/Command.java b/command/src/main/java/com/jagrosh/jdautilities/command/Command.java
index bb014f8..7ba8ccc 100644
--- a/command/src/main/java/com/jagrosh/jdautilities/command/Command.java
+++ b/command/src/main/java/com/jagrosh/jdautilities/command/Command.java
@@ -248,6 +248,27 @@ public abstract class Command
         // availability check
         if(event.getChannelType()==ChannelType.TEXT)
         {
+            //user perms
+            for(Permission p: userPermissions)
+            {
+                if(p.isChannel())
+                {
+                    if(!event.getMember().hasPermission(event.getTextChannel(), p))
+                    {
+                        terminate(event, String.format(userMissingPermMessage, event.getClient().getError(), p.getName(), "channel"));
+                        return;
+                    }
+                }
+                else
+                {
+                    if(!event.getMember().hasPermission(p))
+                    {
+                        terminate(event, String.format(userMissingPermMessage, event.getClient().getError(), p.getName(), "server"));
+                        return;
+                    }
+                }
+            }
+
             // bot perms
             for(Permission p: botPermissions)
             {
@@ -286,27 +307,6 @@ public abstract class Command
                     }
                 }
             }
-            
-            //user perms
-            for(Permission p: userPermissions)
-            {
-                if(p.isChannel())
-                {
-                    if(!event.getMember().hasPermission(event.getTextChannel(), p))
-                    {
-                        terminate(event, String.format(userMissingPermMessage, event.getClient().getError(), p.getName(), "channel"));
-                        return;
-                    }
-                }
-                else
-                {
-                    if(!event.getMember().hasPermission(p))
-                    {
-                        terminate(event, String.format(userMissingPermMessage, event.getClient().getError(), p.getName(), "server"));
-                        return;
-                    }
-                }
-            }
         }
         else if(guildOnly)
         {
-- 
2.23.0

