From 3e1e9757ee94c649e463d0f412c44fb635b9d78e Mon Sep 17 00:00:00 2001
From: Chew <chew@chew.pw>
Date: Tue, 17 Nov 2020 19:26:25 -0600
Subject: [PATCH] Add prefixFunction() in CommandClient


diff --git a/command/src/main/java/com/jagrosh/jdautilities/command/CommandClient.java b/command/src/main/java/com/jagrosh/jdautilities/command/CommandClient.java
index 292754b..f8db009 100644
--- a/command/src/main/java/com/jagrosh/jdautilities/command/CommandClient.java
+++ b/command/src/main/java/com/jagrosh/jdautilities/command/CommandClient.java
@@ -16,6 +16,7 @@
 package com.jagrosh.jdautilities.command;
 
 import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
 
 import java.time.OffsetDateTime;
 import java.util.List;
@@ -91,6 +92,13 @@ public interface CommandClient
      */
     String[] getPrefixes();
 
+    /**
+     * Gets the prefix BiConsumer
+     *
+     * @return A possibly-null prefix BiConsumer
+     */
+    Function<MessageReceivedEvent, String> getPrefixFunction();
+
     /**
      * Returns the visual representation of the bot's prefix. 
      * 
diff --git a/command/src/main/java/com/jagrosh/jdautilities/command/CommandClientBuilder.java b/command/src/main/java/com/jagrosh/jdautilities/command/CommandClientBuilder.java
index 0a5b456..e9f9ce1 100644
--- a/command/src/main/java/com/jagrosh/jdautilities/command/CommandClientBuilder.java
+++ b/command/src/main/java/com/jagrosh/jdautilities/command/CommandClientBuilder.java
@@ -22,8 +22,11 @@ import java.util.function.Consumer;
 import com.jagrosh.jdautilities.command.impl.AnnotatedModuleCompilerImpl;
 import com.jagrosh.jdautilities.command.impl.CommandClientImpl;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Function;
+
 import net.dv8tion.jda.api.OnlineStatus;
 import net.dv8tion.jda.api.entities.Activity;
+import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
 
 /**
  * A simple builder used to create a {@link com.jagrosh.jdautilities.command.impl.CommandClientImpl CommandClientImpl}.
@@ -42,6 +45,7 @@ public class CommandClientBuilder
     private String prefix;
     private String altprefix;
     private String[] prefixes;
+    private Function<MessageReceivedEvent, String> prefixFunction;
     private String serverInvite;
     private String success;
     private String warning;
@@ -69,7 +73,7 @@ public class CommandClientBuilder
      */
     public CommandClient build()
     {
-        CommandClient client = new CommandClientImpl(ownerId, coOwnerIds, prefix, altprefix, prefixes, activity, status, serverInvite,
+        CommandClient client = new CommandClientImpl(ownerId, coOwnerIds, prefix, altprefix, prefixes, prefixFunction, activity, status, serverInvite,
                                                      success, warning, error, carbonKey, botsKey, new ArrayList<>(commands), useHelp,
                                                      shutdownAutomatically, helpConsumer, helpWord, executor, linkedCacheSize, compiler, manager);
         if(listener!=null)
@@ -151,6 +155,22 @@ public class CommandClientBuilder
         this.prefixes = prefixes;
         return this;
     }
+
+    /**
+     * Sets the Prefix Function. Used if you want custom prefixes per server.
+     * <br>Be careful, this function should be quick,
+     * as it's executed every time MessageReceivedEvent is called.
+     * <br>If function returns null, it will be ignored.
+     *
+     * @param prefixFunction
+     *        The prefix function to execute to use
+     *
+     * @return This builder
+     */
+    public CommandClientBuilder setPrefixFunction(Function<MessageReceivedEvent, String> prefixFunction) {
+        this.prefixFunction = prefixFunction;
+        return this;
+    }
     
     /**
      * Sets whether the {@link com.jagrosh.jdautilities.command.CommandClient CommandClient} will use
diff --git a/command/src/main/java/com/jagrosh/jdautilities/command/impl/CommandClientImpl.java b/command/src/main/java/com/jagrosh/jdautilities/command/impl/CommandClientImpl.java
index bc2e503..43c048f 100644
--- a/command/src/main/java/com/jagrosh/jdautilities/command/impl/CommandClientImpl.java
+++ b/command/src/main/java/com/jagrosh/jdautilities/command/impl/CommandClientImpl.java
@@ -74,6 +74,7 @@ public class CommandClientImpl implements CommandClient, EventListener
     private final String prefix;
     private final String altprefix;
     private final String[] prefixes;
+    private final Function<MessageReceivedEvent, String> prefixFunction;
     private final String serverInvite;
     private final HashMap<String, Integer> commandIndex;
     private final ArrayList<Command> commands;
@@ -96,7 +97,7 @@ public class CommandClientImpl implements CommandClient, EventListener
     private CommandListener listener = null;
     private int totalGuilds;
 
-    public CommandClientImpl(String ownerId, String[] coOwnerIds, String prefix, String altprefix, String[] prefixes, Activity activity, OnlineStatus status, String serverInvite,
+    public CommandClientImpl(String ownerId, String[] coOwnerIds, String prefix, String altprefix, String[] prefixes, Function<MessageReceivedEvent, String> prefixFunction, Activity activity, OnlineStatus status, String serverInvite,
                              String success, String warning, String error, String carbonKey, String botsKey, ArrayList<Command> commands,
                              boolean useHelp, boolean shutdownAutomatically, Consumer<CommandEvent> helpConsumer, String helpWord, ScheduledExecutorService executor,
                              int linkedCacheSize, AnnotatedModuleCompiler compiler, GuildSettingsManager manager)
@@ -122,6 +123,7 @@ public class CommandClientImpl implements CommandClient, EventListener
         this.prefix = prefix==null || prefix.isEmpty() ? DEFAULT_PREFIX : prefix;
         this.altprefix = altprefix==null || altprefix.isEmpty() ? null : altprefix;
         this.prefixes = prefixes==null || prefixes.length == 0 ? null : prefixes;
+        this.prefixFunction = prefixFunction;
         this.textPrefix = prefix;
         this.activity = activity;
         this.status = status;
@@ -391,6 +393,12 @@ public class CommandClientImpl implements CommandClient, EventListener
         return prefixes;
     }
 
+    @Override
+    public Function<MessageReceivedEvent, String> getPrefixFunction()
+    {
+        return prefixFunction;
+    }
+
     @Override
     public String getAltPrefix()
     {
@@ -512,6 +520,14 @@ public class CommandClientImpl implements CommandClient, EventListener
             }
         }
         // Check for prefix
+        // Run Function check if there is one, then fallback to normal prefixes
+        if (prefixFunction != null)
+        {
+            String prefix = prefixFunction.apply(event);
+            // Don't lowercase, up to Function to handle this
+            if (prefix != null && rawContent.startsWith(prefix))
+                parts = splitOnPrefixLength(rawContent, prefixFunction.apply(event).length());
+        }
         if(parts == null && rawContent.toLowerCase().startsWith(prefix.toLowerCase()))
             parts = splitOnPrefixLength(rawContent, prefix.length());
         // Check for alternate prefix
-- 
2.29.2

